src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
         end
     end
 
-    pp = similar(p)
+    cp = _compile_permutation!(Base.copymutable(p))
+
+    if fun === Base.invpermute!! 
+        pop!(cp)
+        reverse!(cp)
+        push!(cp, 0)
+    end
 
     for (i, col) in enumerate(eachcol(df))
         if !(i in toskip)
-            copyto!(pp, p)
-            fun(col, pp)
+            _cycle_permute!(col, cp)
         end
     end
     return df
 end
 
+# convert a classical permutation to cycle notation, 
+# destroying the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    Base.require_one_based_indexing(p)
+    out = similar(p, 3length(p)÷2)
```

Comment:
```suggestion
    out = similar(p, 3 * length(p) ÷ 2)
```

Reply:
also maybe add a comment that this length is guaranteed to be enough because we are not recording length 1 cycles.